src/quickunlock/PinUnlock.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+ *  Copyright (C) 2023 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "PinUnlock.h"
+
+#include "crypto/CryptoHash.h"
+#include "crypto/Random.h"
+#include "crypto/SymmetricCipher.h"
+
+#include <QInputDialog>
+#include <QRegularExpression>
+
+#define MIN_PIN_LENGTH 4
+#define MAX_PIN_LENGTH 8
+#define MAX_PIN_ATTEMPTS 3
+
+bool PinUnlock::isAvailable() const
+{
+    return true;
+}
+
+QString PinUnlock::errorString() const
+{
+    return m_error;
+}
+
+bool PinUnlock::setKey(const QUuid& dbUuid, const QByteArray& data)
+{
+    QString pin;
+    QRegularExpression pinRegex("^\\d+$");
+    while (true) {
+        bool ok = false;
+        pin = QInputDialog::getText(
+            nullptr,
+            QObject::tr("Quick Unlock Pin Entry"),
+            QObject::tr("Enter a %1 to %2 digit pin to use for quick unlock:").arg(MIN_PIN_LENGTH).arg(MAX_PIN_LENGTH),
+            QLineEdit::Password,
+            {},
+            &ok);
+
+        if (!ok) {
+            m_error = QObject::tr("Pin setup was canceled. Quick unlock has not been enabled.");
+            return false;
+        }
+
+        // Validate pin criteria
+        if (pin.length() >= MIN_PIN_LENGTH && pin.length() <= MAX_PIN_LENGTH && pinRegex.match(pin).hasMatch()) {
+            break;
+        }
+    }
+
+    // Hash the pin and use it as the key for the encryption
+    CryptoHash hash(CryptoHash::Sha256);
+    hash.addData(pin.toLatin1());
+    auto key = hash.result();
+
+    // Generate a random IV
+    auto iv = Random::instance()->randomArray(SymmetricCipher::defaultIvSize(SymmetricCipher::Aes256_GCM));
+
+    // Encrypt the data using AES-256-CBC
+    SymmetricCipher cipher;
+    if (!cipher.init(SymmetricCipher::Aes256_GCM, SymmetricCipher::Encrypt, key, iv)) {
+        m_error = QObject::tr("Failed to init KeePassXC crypto.");
+        return false;
+    }
+    QByteArray encrypted = data;
+    if (!cipher.finish(encrypted)) {
+        m_error = QObject::tr("Failed to encrypt key data.");
+        return false;
+    }
+
+    // Prepend the IV to the encrypted data
+    encrypted.prepend(iv);
+    // Store the encrypted data and pin attempts
+    m_encryptedKeys.insert(dbUuid, qMakePair(1, encrypted));
```

Comment:
We should probably pad the serialised key to a specific length to avoid leaking length information. I'm in general only half happy with this whole thing, since the PIN attempt count matters only as long as KeePassXC stays in full control of the cipher text (at which point it might as well not be encrypted at all).

On the other hand, the entire database is in memory naked at this point, so none of this really matters probably.

At the very least, this map should be mlocked.

Reply:
Hmmm good idea, should maybe apply Argon2 to the pin to provide protection against brute force

Reply:
I was thinking the same, though with a numerical four-digit PIN, Argon2 would add only very little security.

Reply:
The key is always 32 bytes long due to hashing the pin and (as of now) running it through Argon2. I also increased the minimum pin size to 6 digits.